Do not stack several automatic dialogs

ArturoManzoli opened the pull request:
- Implemented a dialog queue to handle potential modal stack-ups.

ArturoManzoli commented on src/components/InteractionDialog.vue:
```diff
@@ -1,5 +1,11 @@
 <template>
-  <v-dialog v-model="internalShowDialog" :persistent="persistent" :width="maxWidth || 'auto'">
+  <v-dialog
+    v-model="internalShowDialog"
+    :persistent="persistent"
```
The other dialogs will be unchanged and uncoupled from the queue. The ones that will be queued are only those that can potentially stack up, like the ones on the Cockpit startup.

It's also done like this on snackbar queue systems. Whenever a message can stackup, you queue it.

rafaellehmkuhl replied:
I mean that the queue system should be external to the `InteractionDialog.vue` component, otherwise we are putting queue logic on dialogs that should not be queue-aware.

ArturoManzoli replied:
In reality the InteractionDialog is uncoupled from the queue system. 

The queue uses a logic based on dependency injection. If you check on App.vue, the components are loaded on startup and queued independently and the enqueueDialog function takes the component, its props and an id as parameter. 

`interfaceStore.enqueueDialog({
        id: 'VehicleDiscoveryDialog',
        component: VehicleDiscoveryDialog,
        props: {},
      })`

For example, the system update dialog will only be queued if the function that checks for new versions triggers the dialog and will inject the component on the function to do so.
Actually it doesn't have to be a dialog. Could be virtually any other component we have.

rafaellehmkuhl replied:
What I'm saying is that the interaction dialog should not be the one deciding on what happens with the log queue (`interfaceStore.openNextDialogOnQueue()`), but instead there should be an observer outside the interaction dialog checking if there's any interaction dialog open, and if not, open the next dialog in the queue.

The interaction dialog should be what is is, a dialog. It does not need to be aware of the queue. It's an unnecessary architecture coupling.

ArturoManzoli replied:
But the InteractionDialog isn't aware of the queue. The only modification I made to it was the closing logic so it can handle the queue

rafaellehmkuhl replied:
Discussed the details in a call. Main topic was to uncouple to facilitate maintenance, shrink the codebase and allow easy future changes (for new auto-instantiated dialogs).

ArturoManzoli commented on src/App.vue:
```diff
@@ -312,11 +312,13 @@
       </div>
     </v-main>
   </v-app>
-  <About v-if="showAboutDialog" @update:show-about-dialog="showAboutDialog = $event" />
```
Done

rafaellehmkuhl commented on src/views/ConfigurationGeneralView.vue:
```diff
@@ -60,7 +60,7 @@
               class="bg-[#FFFFFF22] mt-3 mb-2 shadow-2"
               variant="flat"
-              @click="showDiscoveryDialog = true"
+              @click="handleVehicleDiscoveryDialog"
```
This is a user-called dialog. What we are doing here is blocking the user to open it until the other dialogs are close?

ArturoManzoli replied:
This dialog will be opened on the startup, automatically by the system, and also by an user command. When opened by the user it will pop up normally

ArturoManzoli replied:
I see what you mean now. I'll make the necessary changes

ArturoManzoli replied:
Done. Now the VehicleConfigurationDialog will open by both ways.

rafaellehmkuhl commented on src/types/interface.ts:
```diff
@@ -0,0 +1,14 @@
+export type DialogOnQueue = {
```
To not confuse it with typescript's `interface`, could we rename this file to `ui.ts`?

rafaellehmkuhl commented on src/App.vue:
```diff
@@ -312,11 +312,13 @@
   <VideoLibraryModal :open-modal="interfaceStore.isVideoLibraryVisible" />
-  <VehicleDiscoveryDialog v-model="showDiscoveryDialog" show-auto-search-option />
-  <UpdateNotification v-if="isElectron()" />
+  <About v-if="showAboutDialog" @update:show-about-dialog="showAboutDialog = $event" />
```
This line is removed from 315 (before the `VideoLibraryModal` component) and re-added on 316 (after the `VideoLibraryModal` component). Why was that??

rafaellehmkuhl commented on src/App.vue:
```diff
@@ -726,17 +730,43 @@ onBeforeUnmount(() => {
 const preventAutoSearch = useStorage('cockpit-prevent-auto-vehicle-discovery-dialog', false)
 
-onMounted(() => {
+const runElectronStartup = async (): Promise<void> => {
+  if (isElectron()) {
+    interfaceStore.enqueueDialog({
+      id: 'UpdateNotification',
+      component: UpdateNotification,
+      props: {},
+    })
+    // Wait 5 seconds to check if we're connected to a vehicle
+    setTimeout(() => {
+      if (vehicleStore.isVehicleOnline) return
+      interfaceStore.enqueueDialog({
+        id: 'VehicleDiscoveryDialog',
+        component: VehicleDiscoveryDialog,
+        props: {},
+      })
+      if (interfaceStore.dialogQueue[0].id === 'VehicleDiscoveryDialog') {
+        interfaceStore.openNextDialogOnQueue()
+      }
+    }, 5000)
+  }
```
Can you move this function to after the `onMounted`? This way git knows its a change on the `onMounted` function and not a complete rewrite of those lines. 

It's important to prevent those changes on the positions of code blocks when not needed.

ArturoManzoli commented:
On last patch:

* Dialog queue now monitors the dom for its open elements by id, and once they cease to exist, they open the next one on the line, decoupling the enqueued elements from the queue logic.
* Reverted changes made on components like UpdateNotification, VehicleDiscoveryDialog and InteractiveDialog;

rafaellehmkuhl commented on src/components/Tutorial.vue:
```diff
@@ -403,6 +392,7 @@ const handleKeydown = (event: KeyboardEvent): void => {
 watch(
   () => interfaceStore.isTutorialVisible,
   (val) => {
+    console.log('🚀 ~ val:', val)
```
This was in an old PR, was asked to be removed, and is being added here again.

rafaellehmkuhl commented on src/components/UpdateNotification.vue:
```diff
@@ -75,7 +79,10 @@ onBeforeMount(() => {
     dialogActions.value = []
     showProgress.value = false
-    showUpdateDialog.value = true
+    if (interfaceStore.activeDialog?.id === 'UpdateNotification' || interfaceStore.activeDialog === undefined) {
+      showUpdateDialog.value = true
+      return
+    }
```
This is still putting the instantiation logic inside the components, just with another name/approach, and there's no need for that.

This is solvable with a much easier approach where the one controlling the issue of automatic notifications (the App.vue file) is the one who instantiates the component, based on the queue logic (show tutorial, when it closes show the discovery dialog, when it closes show the user selection, etc). There's no need for any logic around the automation of the component mounting to be inside the component itself.

Every new dialog component still has to implement logic around an automatic issuing, which it should't need to be aware off. The dialog components should only care about their own content.

rafaellehmkuhl replied:
@ArturoManzoli

ArturoManzoli replied:
I understand your point and that's pretty much what I intended to do with the composables implementation. 
As you can see on the other components, they didn't need any logic inside them. Just the model-value and its emit, that are standard to vue.

Notice that the changes I made to the update component weren't related to queue logic.

The problem I faced with the `UpdateNotification` component is that the logic deciding if it will be opened or not is inside it. So, I couldn't conditionally queue it as I did with the other 2.
And once this component must be loaded on the queue, so it will check for updates and run its own logic to decide if it will pop on the screen or not, it has also to be closed with the emission of the model-value. Otherwise, the queue was just stuck on it no matter if it has updates or no.
The component wasn't being closed after checking, and having no updates.

If we extract the update logic to another file and provide a flag like `cockpitHasUpdates`, we could conditionally add it to the queue and this adjusts would be needed into it

ArturoManzoli commented:
@rafaellehmkuhl, 
I have updated the approach to the dialog queue. Now there is a composable that controls the queue and it only needs the dialogs to have a v-model and emit its update to it on close. 

I think it's a cleaner and more efficient way to solve the dialog stack issue Cockpit have.

I had to slightly modify the opening logic on `UpdateNotification.vue`, besides the regular `props` and  `emit`. 
Since I didn't have the option of conditionally adding this dialog to queue, I added a closing on no update available. This way the component unclogged the queue and life went on normally again :)

rafaellehmkuhl commented on src/components/Tutorial.vue:
```diff
@@ -98,7 +98,18 @@ const { openSnackbar } = useSnackbar()
-const showTutorial = ref(true)
+const props = defineProps<{
+  /**
+   *
```
Missing docs.

rafaellehmkuhl commented on src/components/UpdateNotification.vue:
```diff
@@ -27,17 +27,33 @@
 
+const props = defineProps<{
+  /**
+   *
```
Missing docs

rafaellehmkuhl commented on src/components/UpdateNotification.vue:
```diff
@@ -52,6 +68,23 @@ const updateInfo = ref({
+    if (props.modelValue && updateAvailable.value) {
+      showUpdateDialog.value = true
+    } else {
+      openSnackbar({ message: 'No updates available.', variant: 'success' })
```
This notification is wrong. If there's an update available but the modelValue props is false, it will enter here.

ArturoManzoli commented on src/App.vue:
```diff
@@ -310,10 +310,11 @@
-  <Tutorial v-if="interfaceStore.isTutorialVisible" />
   <VideoLibraryModal v-if="interfaceStore.isVideoLibraryVisible" />
-  <VehicleDiscoveryDialog v-model="showDiscoveryDialog" show-auto-search-option />
-  <UpdateNotification v-if="isElectron()" />
+  <Tutorial :model-value="interfaceStore.isTutorialVisible" />
+  <div v-if="WrappedDialog">
+    <component :is="WrappedDialog" />
+  </div>
```
This is not for the queue. this for the manual opening on the General settings. There is also a similar implementation for the vehicle discovery dialog.
The other way to manually open those, is to create a single item queue for them when manually called.

rafaellehmkuhl replied:
So if we open those manually and they are triggered to be open, will there be two instances of them in the screen?

rafaellehmkuhl replied:
@ArturoManzoli

rafaellehmkuhl replied:
I don't see a problem with this as long as they are not the same dialog.

If the user has a tutorial that was automatically open, it's fine if they open a vehicle discovery one manually, and this manual one gets on top of the other, but if the user has a vehicle discovery manually opened, an automatic one should not be opened over it.

Does it make sense?

rafaellehmkuhl reviewed (changes requested):
There's apparently a bug in the logic.

A tutorial is always opening here, even if I click not to show it again.

https://github.com/user-attachments/assets/d85d4492-b889-4576-97ef-67c7c14549a4

ArturoManzoli commented:
> There's apparently a bug in the logic.
> 
> A tutorial is always opening here, even if I click not to show it again.

Yep, the `don't show again` button behaves correctly on step 1. On other steps it's indeed not saving on the storage.
Fixing that...

rafaellehmkuhl reviewed (changes requested):
The PR is reviving a bug I solved some PRs ago, where the Tutorial is affecting the UI even when not opened.

This is happening because the `v-if` was removed from the Tutorial component.

<img width="1500" alt="image" src="https://github.com/user-attachments/assets/f2e8992b-bb6a-41d7-a2cb-5970e3aa0a1d" />

ArturoManzoli commented on src/components/Tutorial.vue:
```diff
@@ -208,31 +219,27 @@ const handleStepChange = (newStep: number): void => {
       interfaceStore.componentToHighlight = 'menu-trigger'
       interfaceStore.currentSubMenuComponentName = null
       interfaceStore.currentSubMenuName = null
-      interfaceStore.userHasSeenTutorial = false
```
Actually this variable is used on Cockpit's boot to decide if the tutorial will be auto opened or not. On step change, the tutorial does not close and doesn't need this variable set anyway.

It was removed because it was causing the tutorial to always show on boot if it was closed on a step different than step zero. This variable should only be modified by the `Don't show again` or `Show on startup` buttons.

rafaellehmkuhl replied:
But if the user has interacted with the tutorial, it means they are still on it and should indeed be auto-opened.

The bug I mentioned was not on master, because once you click the "don't show again" button and restarts, it wouldn't open again. That was introduced here. It's not on master.

rafaellehmkuhl replied:
Behavior on master:


https://github.com/user-attachments/assets/10940d6f-b159-4b38-a26f-c9ac10badf44

rafaellehmkuhl replied:
Give the following situation:
- User has seen the tutorial before (has-seen-tutorial === true)
- User clicks to open the tutorial explicitly
- User starts interacting with the tutorial (clicks next step)
- User or system itself restarts

Does it make sense for the tutorial not to reopen, and in the exactly step the user was in before the restart?

I don't think it is such a feature that is worth that much discussion over, but my point in the end is: if this is working on master and the logic makes sense, why is it being changed here?

ArturoManzoli replied:
I agree that this takes time from more important things :)

The feature that returned into the same step after a reboot was added originally only on step 5, that was when the user was being presented to the vehicle address configuration. On this tutorial step the general configuration page was opened and he was invited to input its vehicle netw. location. After that, cockpit reboots automatically and it was important to put the user exactly on the same spot he was before the mandatory reboot. That was the only need for the 'tutorial resume' function.
After that, some weeks ago, the tutorial was refactored to include the changes needed to handle the tools submenus, and during that it was added this resume feature to all of the steps. But this messes up with the original intention and with a variable that wasn't used for that.
This was causing some misfires from that variable, so I decided to set it back to the use it had before the refactoring.

Nothing really special about that, this was all to explain why I decided to delete those lines. Now the component works fine with the new modifications added (v-model and emits).

rafaellehmkuhl replied:
But the logic of reopening if the user interacted with it makes sense, that's why it was changed on that PR.

ArturoManzoli commented on src/components/Tutorial.vue:
```diff
@@ -332,7 +333,6 @@ const alwaysShowTutorialOnStartup = (): void => {
 
 const nextTutorialStep = (): void => {
   if (currentTutorialStep.value === steps.length) {
-    dontShowTutorialAgain()
```
Ok, this one makes sense :)

ArturoManzoli commented:
After a meeting it was decided to keep the tutorial as close as possible to its last refactoring, adapting it to use the `v-model` and `emit` triggerings, also taking care to prevent css/highlighting leaks after the component is closed.

rafaellehmkuhl reviewed (changes requested):
I've tried to use the tutorial with the PR and the following is happening:

https://github.com/user-attachments/assets/e4fb2f14-7d0a-4037-8933-5eaec8ffbdcc

I cannot reproduce the same behavior on master thought.

It happens consistently if you top in the "General configuration" step, refresh Cockpit, and click the buttons in the following steps: "next, previous, previous, next". It doesn't happen only in this situation. It is just an example to make it easier to reproduce the problem.

ArturoManzoli commented:
> I've tried to use the tutorial with the PR and the following is happening:
> 
> I cannot reproduce the same behavior on master thought.

I was able to reproduce the bug you described on the PR's branch, and also couldn't reproduce on master.
I compared the main components that are involved on the tutorial operation (Tutorial.vue, App.vue and appInterface.ts) and there are no significant changes between the two branches on those files. There are some small changes relative to this and to other tasks, but none explains the almost instant closing of the main menu and its subcomponents on the tutorial's command to open them.
I suspect that the problem is relative to the `submenuComponentName` going from `null` to some content, but this doesnt explain much.

Since this is a edge case and doesn't affect Cockpit directly, nor the dialog queue, I suggest we explore this bug with more care in a separate issue.

ArturoManzoli commented:
> It happens consistently if you top in the "General configuration" step, refresh Cockpit, and click the buttons in the following steps: "next, previous, previous, next". It doesn't happen only in this situation. It is just an example to make it easier to reproduce the problem.

It was (weirdly) related to the submenu component being null. Changing it to undefined and adding a None submenu component did solve the issue.

Please try it out now.

rafaellehmkuhl reviewed (changes requested):
The bug is still happening for me, the same way as before:

https://github.com/user-attachments/assets/8915e151-665d-479e-be1d-c7d6d5aa8940

Checked again doing hard-refreshes and restarting the Vite server, for both master and this branch, and the bug is consistent.
